### PR TITLE
feat: Expose execution control from Rust to Python

### PR DESCRIPTION
**Pull request by bjchambers**
This is a building block for allowing Python code to expose batches as generators, async generators, etc.

### CODE REVIEW

File: sparrow-py/src/execution.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+impl Execution {
+    pub(crate) fn new(execution: RustExecution) -> Self {
+        Self(Some(execution))
+    }
+}
+
+#[pymethods]
+impl Execution {
+    fn collect_pyarrow(&mut self, py: Python<'_>) -> Result<Vec<PyObject>, Error> {
```
Comment (kerinin) [inline sparrow-py/src/execution.rs:19]: Could the be confused with `Expr.collect()`?

File: crates/sparrow-session/src/execution.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            }
+        }
+    }
+
+    pub async fn next(&mut self) -> error_stack::Result<Option<RecordBatch>, Error> {
+        self.is_done()?;
+        Ok(self.output.next().await)
+    }
+
+    pub async fn next_blocking(&mut self) -> error_stack::Result<Option<RecordBatch>, Error> {
```
Comment (kerinin) [inline crates/sparrow-session/src/execution.rs:78]: `collect_all_blocking` isn't `async` but this is - intentional?

**Review by kerinin — APPROVED** [review]
So I guess the eventual idea is to wrap the `async` rust functions with async Python functions?

File: sparrow-py/src/execution.rs (hunk 1 above)
Comment (bjchambers) [reply sparrow-py/src/execution.rs:19]: I don't think so? This is part of `sparrow_py._ffi.Execution`, while that is a method on `sparrow-py.expr.Expr`, and has a different name (`collect` vs. `collect_pyarrow`).

File: crates/sparrow-session/src/execution.rs (hunk 2 above)
Comment (bjchambers) [reply crates/sparrow-session/src/execution.rs:78]: Oh. This should be non-async (it uses the tokio `runtime` directly, and waits for that).

**Comment by bjchambers** [comment]
> So I guess the eventual idea is to wrap the `async` rust functions with async Python functions?

Yeah. I'm not sure how that will work -- I think we need to have the rust runtime (tokio) run the async function and then call a method to resolve the Python future on it's pool. But not 100% sure (exploring next).

File: sparrow-py/src/execution.rs (hunk 1 above)
Comment (kerinin) [reply sparrow-py/src/execution.rs:19]: It's not a big deal, but in one case we're describing an operation and in another, we're controlling execution. I could imagine the execution class moving around or becoming more "public". Not a big deal - I was just wondering if there's a similar word we could use.

Comment (bjchambers) [reply sparrow-py/src/execution.rs:19]: Ah. I think it's pretty common for the "collect the results of this thing"... I think we're probably OK -- one being called `collect` and the other `collect_pyarrow` or `collect_dataframes` etc. In terms of the user, it's about whether they've already run or not:

```
collect().run()
vs.
run().collect_dataframes()
```

**Pull request merged**